### PR TITLE
Reorganizes project for zipkin-server integration as io.zipkin.gcp

### LINKED ISSUE #45: Proposal: zipkin-gcp
There are a couple of zipkin subprojects targeting different cloud environments. We should look into how to evolve this into the same fold, e.g., zipkin-gcp.  zipkin-gcp should can have the following:

- [ ] Senders
  - [ ] Pub/Sub (sender-pubsub)
  - [ ] Stackdriver Trace (sender-stackdriver)
- [ ] Storage
  - [ ] Stackdriver Storage (storage-stackdriver)
- [ ] Collectors
  - [ ] Pub/Sub (collector-pubsub)
- [ ] Autoconfigurers

We can also use `spring-cloud-gcp-core` as the basis for getting Project ID and Credentials.

Current `stackdriver-zipkin` behavior can ultimately be composed by:
- storage-stackdriver
- spring-cloud-gcp-core
- autoconfigurers

### PR DESCRIPTION
**Pull request by adriancole**
This reorganizes and refactors such that this can be a normal part of
OpenZipkin. Notable adjustments include:
* switching to default OpenZipkin template (format, build, publishing)
* refactors the `StorageComponent` to have less dependencies
  * this allows it to be re-used as a span reporter in android
* updates to latest grpc (just to make sure it works)
* reorganizes project such that it builds a module plug-in
  * this allows weaving in stackdriver with kafka for example

Fixes #45

### CODE REVIEW

**Comment by adriancole** [comment]
note: I haven't ported the metrics/executor code at the moment, mainly as it is uncommon with other storage providers, but also I'd like to see if it remains an issue since we are using grpc directly now.

There are also some other content I've not moved.. feel free to add commits if you like.

One thing you'll notice is we don't publish docker from this repo. Once in openzipkin, we'd make docker-zipkin-gcp to handle that, producing a docker image called openzipkin/zipkin-gcp which includes re-assigned support for stackdriver

File: storage-stackdriver/src/main/java/zipkin2/storage/stackdriver/StackdriverSpanConsumer.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin2.storage.stackdriver;
+
+import com.google.devtools.cloudtrace.v1.PatchTracesRequest;
+import com.google.devtools.cloudtrace.v1.Trace;
+import com.google.devtools.cloudtrace.v1.Traces;
+import com.google.protobuf.Empty;
+import io.grpc.CallOptions;
```
Comment (adriancole) [inline storage-stackdriver/src/main/java/zipkin2/storage/stackdriver/StackdriverSpanConsumer.java]: you'll notice this doesn't pull in code like oauth libraries and netty. We might consider the former, but the latter makes it really version sensitive with the server (later this can be simply wrapped to create a zipkin reporter folks can put in apps.. might make sense to pre-ordain a zipkin-stackdriver-netty variant for those who can accept an externally provided netty+boringssl combo)

File: storage-stackdriver/src/main/java/zipkin2/storage/stackdriver/StackdriverStorage.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  abstract String projectId();
+
+  abstract ManagedChannel channel();
+
+  abstract CallOptions callOptions();
+
+  abstract boolean shutdownChannelOnClose();
+
+  @Override public SpanStore spanStore() {
+    throw new UnsupportedOperationException("Read operations are not supported");
```
Comment (adriancole) [inline storage-stackdriver/src/main/java/zipkin2/storage/stackdriver/StackdriverStorage.java]: not yet supported... later I'll probably help make at least get by trace ID supported as folks have asked for this if only to test if their trace got there.

File: storage-stackdriver/src/main/java/zipkin2/storage/stackdriver/StackdriverStorage.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  @Override public SpanConsumer spanConsumer() {
+    return new StackdriverSpanConsumer(channel(), projectId(), callOptions());
+  }
+
+  @Override
+  public CheckResult check() {
+    return CheckResult.OK;
```
Comment (adriancole) [inline storage-stackdriver/src/main/java/zipkin2/storage/stackdriver/StackdriverStorage.java]: this is an awful health check and makes some things very hard to figure out. What is a cheap health check we can make given a channel and knowing we are connected to stackdriver?

File: pom.xml [hunk 4]
```diff
@@ -1,161 +1,364 @@
...
+    <module>autoconfigure</module>
+    <module>storage-stackdriver</module>
+  </modules>
+
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <project.build.resourceEncoding>UTF-8</project.build.resourceEncoding>
+
+    <!-- default bytecode version for src/main -->
+    <main.java.version>1.6</main.java.version>
```
Comment (adriancole) [inline pom.xml]: ps not sure if grpc (for android) goes down to this level. if 1.7, we should adjust

File: pom.xml [hunk 5]
```diff
@@ -1,161 +1,364 @@
...
+    <!-- only used for protos, we could possibly obviate this if a problem -->
+    <grpc-google-devtools-cloudtrace-v1.version>0.1.1</grpc-google-devtools-cloudtrace-v1.version>
+
+    <!-- must match (ex. use spring-boot used by zipkin-server) -->
+    <zipkin-reporter2.version>2.1.4</zipkin-reporter2.version>
+    <zipkin.version>2.2.3-SNAPSHOT</zipkin.version>
+    <spring-boot.version>1.5.8.RELEASE</spring-boot.version>
+
+    <!-- must match https://github.com/grpc/grpc-java/blob/master/SECURITY.md#netty -->
+    <grpc.version>1.7.0</grpc.version>
```
Comment (adriancole) [inline pom.xml]: thanks tons to @saturnism for writing that section. saved my butt

File: pom.xml [hunk 6]
```diff
@@ -1,161 +1,364 @@
...
+
+    <license-maven-plugin.version>2.11</license-maven-plugin.version>
+
+    <slf4j.version>1.7.25</slf4j.version>
+    <!-- only used for protos, we could possibly obviate this if a problem -->
+    <grpc-google-devtools-cloudtrace-v1.version>0.1.1</grpc-google-devtools-cloudtrace-v1.version>
+
+    <!-- must match (ex. use spring-boot used by zipkin-server) -->
+    <zipkin-reporter2.version>2.1.4</zipkin-reporter2.version>
+    <zipkin.version>2.2.3-SNAPSHOT</zipkin.version>
```
Comment (adriancole) [inline pom.xml]: awaiting a netty-free zipkin release (as we are overriding netty and boringssl to match grpc)

File: pom.xml [hunk 7]
```diff
@@ -1,161 +1,364 @@
...
+          <tagNameFormat>@{project.version}</tagNameFormat>
+        </configuration>
+      </plugin>
+
+      <plugin>
+        <groupId>io.zipkin.centralsync-maven-plugin</groupId>
+        <artifactId>centralsync-maven-plugin</artifactId>
+        <version>0.1.0</version>
+        <configuration>
+          <packageName>zipkin-gcp</packageName>
```
Comment (adriancole) [inline pom.xml]: once we get closer to the ball, I'll create zipkin-gcp in bintray and circleci; folks who want to help with releases and/or build support can be added to both. The release is tag-driven and automatic.

File: pom.xml [hunk 8]
```diff
@@ -1,161 +1,364 @@
...
+          <!-- Creates source jar -->
           <plugin>
-            <groupId>org.apache.maven.plugins</groupId>
-            <artifactId>maven-gpg-plugin</artifactId>
-            <version>1.6</version>
+            <artifactId>maven-source-plugin</artifactId>
+            <version>3.0.1</version>
             <executions>
               <execution>
-                <id>sign-artifacts</id>
```
Comment (adriancole) [inline pom.xml]: we use bintray's signatures in openzipkin (preventing us from having to coordinate keys)

File: pom.xml [hunk 9]
```diff
@@ -1,161 +1,364 @@
...
               </execution>
             </executions>
           </plugin>
+
+          <plugin>
+            <artifactId>maven-javadoc-plugin</artifactId>
+            <version>2.10.4</version>
+            <configuration>
+              <failOnError>false</failOnError>
+              <excludePackageNames>zipkin.aws.internal,zipkin.aws.internal.*
```
Comment (adriancole) [inline pom.xml]: whoops.. guess you know where find/replace came from...

**Comment by adriancole** [comment]
instead of adding custom backlog stats, I've redone the base proxy server so that it can swallow 3x more requests with no TCP errors https://github.com/openzipkin/zipkin/pull/1806#issuecomment-346387611

tl;dr; I'd like to move towards merge of this codebase so that we can integrate it.

**Comment by adriancole** [comment]
@bogdandrutu ps this should be ready to go, imho. I did notice some stackdriver v2 action. If that's stable, I don't mind switching to that. Otherwise, let's get ready to do the org move?

**Comment by anuraaga** [comment]
Just wondering, would we want do do one more release before the big move? Looking forward to the new version of the translator :)

**Comment by adriancole** [comment]
yep. that would be the plan. to do a release here, then move, merge,
release there

**Comment by BrendanBall** [comment]
@anuraaga @bogdandrutu if you guys haven't gone on holiday yet, how much work is it to do the new release with zipkin 2.4? Otherwise I guess we'll wait till next year.

**Comment by bogdandrutu** [comment]
@adriancole I am not sure I completely understand the plan. Do we need a release of the current repo, then a release with this PR merged?

**Comment by bogdandrutu** [comment]
@BrendanBall I am on vacation but the work to do the release is not that huge so I may be able to do it after I understand the plan.

**Comment by adriancole** [comment]
So, the plan is pretty simple:

* release current master (you have to as you have the keys)
* add me as owner so I can move the repo (I think any owner can do this)
* I immediately merge and release this PR (I will deal with release setup,
but if it delays much longer can't as I have holiday coming up)

**Comment by adriancole** [comment]
guessing this just needs to wait till post holiday. ping me when ready and I'll update this to latest versions which have increased since opened 2 months ago

**Comment by bogdandrutu** [comment]
can you rebase this PR?

**Comment by adriancole** [comment]
rebased

File: storage-stackdriver/src/main/java/zipkin2/stackdriver/translation/TraceTranslator.java [hunk 10]
```diff
@@ -66,7 +52,7 @@ public TraceTranslator(String projectId) {
         finishTrace(currentTrace, translatedTraces);
 
         currentTrace = Trace.newBuilder();
-        currentTrace.setProjectId(this.projectId);
+        currentTrace.setProjectId(projectId);
         currentTrace.setTraceId(traceId);
```
Comment (saturnism) [inline storage-stackdriver/src/main/java/zipkin2/stackdriver/translation/TraceTranslator.java]: is there any chance a 128bit ID results in < 32 digits of Hex? I.e., do we left pad with 0's like https://github.com/GoogleCloudPlatform/cloud-trace-java/blob/b39dd53117a315f9fea3ff3508dc4805621c0c80/sdk/core/src/main/java/com/google/cloud/trace/core/TraceId.java#L100

File: storage-stackdriver/src/main/java/zipkin2/stackdriver/translation/SpanTranslator.java [hunk 11]
```diff
@@ -137,7 +136,7 @@ private SpanKind getSpanKind(@Nullable Span.Kind zipkinKind) {
     return SpanKind.UNRECOGNIZED;
   }
 
-  private Timestamp createTimestamp(long microseconds) {
+  private static Timestamp createTimestamp(long microseconds) {
```
[the remark this answers is not in the document]
Comment (saturnism) [reply storage-stackdriver/src/main/java/zipkin2/stackdriver/translation/SpanTranslator.java]: oops, nevermind, i was thinking of sleuth span. scratch this comment.

**Comment by adriancole** [comment]
manually retested and works (integrated manually as we don't have a docker repo yet)

```bash
GOOGLE_APPLICATION_CREDENTIALS=~/Downloads/not-telling.json STORAGE_TYPE=stackdriver PROJECT_ID=zipkin-demo     java     -Dloader.path='stackdriver.jar,stackdriver.jar!/lib'     -Dspring.profiles.active=stackdriver     -cp zipkin.jar     org.springframework.boot.loader.PropertiesLauncher
```

screen shot 2018-02-16 at 10 03 06 am

**Comment by adriancole** [comment]
Thanks for the feedback ray. In general about public, we dont make
something public until it has a use case to be public. Otherwise we make it
possible to break other people's code. In many cases where something is
accidentally public (like zipkin server classes) we have found someone
relying on something we need to change.

Considering the reporter functionality will not need this public, and
neither will tests.. I'm inclined to not expose it.

Why do you need this public?

**Comment by adriancole** [comment]
I am out today but will sort out the Long object thing and also backfill a
test that shows what happens with 64bit trace ID.

**Comment by saturnism** [comment]
@adriancole SpanTranslator was handy to use to translate just a single Span rather than a whole list. But I think that problem goes away once we have a StackdriverSender in the future.

**Comment by adriancole** [comment]
>
> @adriancole <https://github.com/adriancole> SpanTranslator was handy to
> use to translate just a single Span rather than a whole list. But I think
> that problem goes away once we have a StackdriverSender in the future.
>
yep agree tactically it made sense, but since this is a new group id etc.
let's try the sender thing here (and copy/paste internal until that's
published). There's some games the list translater plays in attempts to
monkey with span IDs. I don't think they are always effective (ex client
and server spans are almost never in the same http message), but anyway it
let's us rejig w/o api breaking.

>

File: storage-stackdriver/src/main/java/zipkin2/stackdriver/translation/SpanTranslator.java [hunk 12]
```diff
@@ -110,14 +109,14 @@ private void rewriteIds(Span zipkinSpan, TraceSpan.Builder builder, SpanKind kin
...
-  private long parseUnsignedLong(String id) {
+  private static long parseUnsignedLong(String id) {
     if (id == null) {
       return 0;
     }
     return UnsignedLongs.parseUnsignedLong(id, 16);
   }
 
-  private long rewriteId(Long id) {
+  private static long rewriteId(Long id) {
```
[the remark this answers is not in the document]
Comment (adriancole) [reply storage-stackdriver/src/main/java/zipkin2/stackdriver/translation/SpanTranslator.java]: done

File: storage-stackdriver/src/main/java/zipkin2/stackdriver/translation/SpanTranslator.java [hunk 13]
```diff
@@ -55,21 +51,23 @@ public SpanTranslator() {
...
-  public TraceSpan translate(Span zipkinSpan) {
+  static TraceSpan translate(Span zipkinSpan) {
     TraceSpan.Builder builder = TraceSpan.newBuilder();
     translate(builder, zipkinSpan);
     return builder.build();
   }
 
-  TraceSpan.Builder translate(TraceSpan.Builder spanBuilder, Span zipkinSpan) {
+  static TraceSpan.Builder translate(TraceSpan.Builder spanBuilder, Span zipkinSpan) {
     spanBuilder.setName(zipkinSpan.name());
```
[the remark this answers is not in the document]
Comment (adriancole) [reply storage-stackdriver/src/main/java/zipkin2/stackdriver/translation/SpanTranslator.java]: done

File: storage-stackdriver/src/main/java/zipkin2/stackdriver/translation/TraceTranslator.java (hunk 10 above)
Comment (adriancole) [reply storage-stackdriver/src/main/java/zipkin2/stackdriver/translation/TraceTranslator.java]: nope zipkin2.Span's builder normalizes the IDs (via Span.normalizeTraceId). good Q!
